nomad/node_endpoint.go

```diff
@@ -1595,6 +1622,246 @@ func (n *Node) DeriveVaultToken(args *structs.DeriveVaultTokenRequest,
+						ClusterID: clusterID,
+						AllocID:   alloc.ID,
+						TaskName:  task,
+					}
+					secret, err := n.srv.consulACLs.CreateToken(ctx, sii)
+					if err != nil {
+						return err
+					}
+					results[task] = secret
+				case <-ctx.Done():
+					return nil
+				}
+			}
+		})
+	}
+
+	// Send the input
+	go func() {
+		defer close(input)
+		for _, task := range args.Tasks {
+			select {
+			case <-ctx.Done():
+				return
+			case input <- task:
+			}
+		}
+	}()
+
+	// Wait for everything to complete or for an error
+	createErr := g.Wait()
+
+	accessors := make([]*structs.SITokenAccessor, 0, len(results))
+	tokens := make(map[string]string, len(results))
+	for task, secret := range results {
+		tokens[task] = secret.SecretID
+		accessor := &structs.SITokenAccessor{
+			NodeID:     alloc.NodeID,
+			AllocID:    alloc.ID,
+			TaskName:   task,
+			AccessorID: secret.AccessorID,
+		}
+		accessors = append(accessors, accessor)
+	}
+
+	// If there was an error, revoke all created tokens. These tokens have not
+	// yet been committed to the persistent store.
+	if createErr != nil {
+		n.logger.Error("Consul Service Identity token creation for alloc failed", "alloc_id", alloc.ID, "error", createErr)
+		_ = n.srv.consulACLs.RevokeTokens(context.Background(), accessors, false)
+
+		if recoverable, ok := createErr.(*structs.RecoverableError); ok {
+			reply.Error = recoverable
+		} else {
+			reply.Error = structs.NewRecoverableError(createErr, false).(*structs.RecoverableError)
+		}
+
+		return nil
+	}
+
+	// Commit the derived tokens to raft before returning them
+	requested := structs.SITokenAccessorsRequest{Accessors: accessors}
+	_, index, err := n.srv.raftApply(structs.ServiceIdentityAccessorRegisterRequestType, &requested)
+	if err != nil {
+		n.logger.Error("registering Service Identity token accessors for alloc failed", "alloc_id", alloc.ID, "error", err)
+
+		// Determine if we can recover from the error
+		retry := false
+		switch err {
+		case raft.ErrNotLeader, raft.ErrLeadershipLost, raft.ErrRaftShutdown, raft.ErrEnqueueTimeout:
+			retry = true
+		}
+		setError(err, retry)
+		return nil
+	}
+
+	// We made it! Now we can set the reply.
+	reply.Index = index
+	reply.Tokens = tokens
+	n.srv.setQueryMeta(&reply.QueryMeta)
+	return nil
+}
+
+func tgUsesConnect(tg *structs.TaskGroup) bool {
+	for _, service := range tg.Services {
+		if service.Connect != nil {
+			if service.Connect.Native || service.Connect.SidecarService != nil {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func tasksNotUsingConnect(tg *structs.TaskGroup, tasks []string) []string {
+	var unneeded []string
+	for _, task := range tasks {
+		tgTask := tg.LookupTask(task)
+		if !taskUsesConnect(tgTask) {
+			unneeded = append(unneeded, task)
+		}
+	}
+	return unneeded
+}
+
+func taskUsesConnect(task *structs.Task) bool {
+	if task == nil {
+		// not even in the task group
+		return false
+	}
+	// todo(shoenig): TBD what Kind does a native task have?
```

Comment:
We will probably have to designate an additional kind for native tasks that users must set if they want Nomad to derive an SI token for them. We only identify connect native at the group service level today.